### PR TITLE
Multi Target Delete Support

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR is an enhancement on earlier PR to add support for multi-target delete.

Example: `delete u, m from user u, music m on u.col = m.col`
The SQL here deletes rows from two tables using the join condition.

## Related Issue(s)

- Closes https://github.com/vitessio/vitess/issues/5392
- #14855 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/evalengine/expr_tuple_bvar.go [hunk 1]
```diff
@@ -32,12 +32,6 @@ type (
 		Index     int
 		Type      sqltypes.Type
 		Collation collations.ID
-
-		// dynamicTypeOffset is set when the type of this bind variable cannot be calculated
-		// at translation time. Since expressions with dynamic types cannot be compiled ahead of time,
-		// compilation will be delayed until the expression is first executed with the bind variables
-		// sent by the user. See: UntypedExpr
-		dynamicTypeOffset int
```
Comment (GuptaManan100) [inline go/vt/vtgate/evalengine/expr_tuple_bvar.go:40]: Why are we getting rid of this?

File: go/vt/vtgate/planbuilder/operators/delete.go [hunk 2]
```diff
@@ -144,18 +132,78 @@ func deleteWithInputPlanningForFk(ctx *plancontext.PlanningContext, del *sqlpars
...
+		aVdx, bVdx := getFirstVindex(a.vTbl), getFirstVindex(b.vTbl)
+
+		// Sort nil Vindexes to the start
+		if aVdx == nil || bVdx == nil {
+			return aVdx != nil // true if bVdx is nil and aVdx is not nil
+		}
+
+		// Among non-nil Vindexes, those that need VCursor come first
+		return aVdx.NeedsVCursor() && !bVdx.NeedsVCursor()
+	})
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operators/delete.go:156]: Why is this required?

**Review by GuptaManan100 — APPROVED** [review]
Just a couple of questions, rest lgtm

File: go/vt/vtgate/evalengine/expr_tuple_bvar.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/evalengine/expr_tuple_bvar.go:40]: It is not used in this evalengine type

File: go/vt/vtgate/planbuilder/operators/delete.go (hunk 2 above)
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/delete.go:156]: We need to delete the tables based on the sharding vindex type.
Example
```
user - id is a sharding key, phone as lookup key)
directory - phone as lookup sharding key)
```

`delete u, d from user u join directory d on u.phone = d.phone`

Here if we delete rows from `user` table first then data will be delete from `phone` lookup table as well.
When we try to delete from `directory` table we will not be able to find the data in `phone` lookup table as it was already delete when deleting from `user` table. The delete query for `directory` will not be sent to any shard.

Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/operators/delete.go:156]: Okay, I understand, might be worth adding a comment explaining this.

**Review by frouioui — APPROVED** [review]
